sql/core/src/main/scala/org/apache/spark/sql/execution/dynamicpruning/PartitionPruning.scala

```diff
@@ -247,13 +246,15 @@ object PartitionPruning extends Rule[LogicalPlan] with PredicateHelper with Join
             // otherwise the pruning will not trigger
             var partScan = getPartitionTableScan(l, left)
             if (partScan.isDefined && canPruneLeft(joinType) &&
-                hasPartitionPruningFilter(right)) {
+                hasPartitionPruningFilter(right) &&
+                (canBroadcastBySize(right, conf) || hintToBroadcastRight(hint))) {
               newLeft = insertPredicate(l, newLeft, r, right, rightKeys, partScan.get,
                 canBuildBroadcastRight(joinType))
             } else {
               partScan = getPartitionTableScan(r, right)
               if (partScan.isDefined && canPruneRight(joinType) &&
-                  hasPartitionPruningFilter(left) ) {
+                  hasPartitionPruningFilter(left) &&
+                  (canBroadcastBySize(left, conf) || hintToBroadcastLeft(hint))) {
```

Comment:
@wangyum we should update this place. This is not a pure shortcut, as it changes the behavior and skips DPP under some cases, https://github.com/apache/spark/pull/31984/files#r607540865

Reply:
The first commit is to prevent this case: https://github.com/apache/spark/pull/31984/commits/b510d7da21f8a92af69b1485b72aef6ad5901448

Reply:
That is more like a hack. If we need a shortcut here, we should closely follow the logic of `insertPredicate` and answer this question: when can we safely skip DPP here?

Reply:
If `broadcastOnly=true`, we can't skip DPP here and need to wait until the planner phase to see if it's a broadcast join.

If `broadcastOnly=false`, and `hasBenefits=true`, we always do DPP (this PR changes this behavior). If `hasBenefits=false`, we can't skip DPP here and need to wait until the planner phase to see if it's a broadcast join.

That said, the shortcut here is to skip DPP earlier if we know for sure that the join can't be broadcast join. I feel it's more robust to wait for the planner and see if it' a broadcast join. Do we really need the shortcut?

Reply:
Revert this pr?

Reply:
Let me revert it first. "broadcast hint" is still a valid case but that might be too minor. We need to think more about how to improve the cost model of DPP.